dashboard/app/jobs.go

```diff
@@ -455,6 +546,82 @@ func createJobResp(c context.Context, job *Job, jobKey *db.Key) (*dashapi.JobPol
 	return resp, false, nil
 }
 
+// It would be easier to just check if the User field is empty, but let's also not
+// miss the situation when some actual user sends a patch testing request without
+// patch.
+func isRetestReproJob(job *Job, build *Build) bool {
+	if job.Type != JobTestPatch || job.Patch != 0 {
+		return false
+	}
+	if job.KernelRepo != build.KernelRepo ||
+		job.KernelBranch != build.KernelBranch {
+		return false
+	}
+	return true
+}
+
+func handleRetestedRepro(c context.Context, now time.Time, job *Job, jobKey *db.Key, crashTitle string) error {
+	bugKey := jobKey.Parent()
+	crashKey := db.NewKey(c, "Crash", "", job.CrashID, bugKey)
+	crash := new(Crash)
+	if err := db.Get(c, crashKey, crash); err != nil {
+		return fmt.Errorf("failed to get crash: %v", crashKey)
+	}
+	build, err := loadBuild(c, job.Namespace, crash.BuildID)
+	if err != nil {
+		return fmt.Errorf("failed to query build: %w", err)
+	}
+	if !isRetestReproJob(job, build) {
+		return nil
+	}
+	bug := new(Bug)
+	if err := db.Get(c, bugKey, bug); err != nil {
+		return fmt.Errorf("failed to get bug: %v", bugKey)
+	}
+	// Update the crash.
+	crash.LastStaleCheck = now
+	if crashTitle == "" {
+		// If there was any crash at all, the repro is still not worth discarding.
+		crash.StaleRepro = true
+	}
+	crash.UpdateReportingPriority(build, bug)
+	if _, err := db.Put(c, crashKey, crash); err != nil {
+		return fmt.Errorf("failed to put crash: %v", err)
+	}
+	reproCrashes, crashKeys, err := queryCrashesForBug(c, bugKey, 2)
+	if err != nil {
+		return fmt.Errorf("failed to fetch crashes with repro: %v", err)
+	}
+	// Now we can update the bug.
+	bug.ReproLevel = ReproLevelNone
+	for id, bestCrash := range reproCrashes {
+		if crashKeys[id].Equal(crashKey) {
+			// In Datastore, we don't see previous writes in a transaction...
+			bestCrash = crash
+		}
+		if bestCrash.StaleRepro {
+			continue
+		}
+		if bestCrash.ReproC > 0 {
+			bug.ReproLevel = ReproLevelC
+		} else if bug.ReproLevel != ReproLevelC && bestCrash.ReproSyz > 0 {
+			bug.ReproLevel = ReproLevelSyz
+		}
+	}
+	if crashTitle != "" {
+		// We don't want to confuse users, so only update LastTime if the generated crash
+		// really relates to the existing bug.
+		sameBug := bug.Title == crashTitle || stringInList(bug.AltTitles, crashTitle)
```

Comment:
I think bug.Title must always be present in bug.AltTitles.
However, ideally we also check the current crash AltTitles against all bug.AltTitles, we can use AltTitles to rename crashes over time.

Reply:
That title was taken from the job response, and there's only one title, no alt titles unfortunately.

Reply:
Yes, we did not use it, we should add it. Not sure if it's worth it, though.
But sometimes we do something like de4e4f4d8b58, which works based on alt titles.